src/udev/udev-node.c

```diff
@@ -232,7 +228,39 @@ static int link_update(sd_device *dev, const char *slink, bool add) {
                                 r = -errno;
                 } while (r == -ENOENT);
 
-        return r;
+        for (i = 0; i < LINK_UPDATE_MAX_RETRIES; i++) {
+                struct stat st1 = {}, st2 = {};
+
```

Comment:
i figure performance of the contended case can be improved by inserting a line like this:

```
#define CONTENTION_SLEEP_MAX_USEC (500 * USEC_PER_MSEC)
…
(void) usleep(random_u64() % (CONTENTION_SLEEP_MAX_USEC * i / (LINK_UPDATE_MAX_RETRIES-1));
```

i.e. sleeping by randomly increasing times, to make it less likely we collide with others.

Reply:
I don't think this is necessary unless scaling proves to be a problem in practice.

Reply:
you mentioned you reach the loop limit all the time? I'd call that a problem...